src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -342,26 +343,19 @@ private boolean hasMateReferenceName() {
      * the reference index to NO_ALIGNMENT_REFERENCE_INDEX.
      *
      * @param referenceName - must not be null
+     * @throws IllegalArgumentException if {@code referenceName} is null
      */
     public void setReferenceName(final String referenceName) {
         if (null == referenceName) {
             throw new IllegalArgumentException(
                     "Reference name must not be null. Use SAMRecord.NO_ALIGNMENT_REFERENCE_NAME to reset the reference name.");
         }
-        else if (NO_ALIGNMENT_REFERENCE_NAME.equals(referenceName)) {
-            mReferenceName = NO_ALIGNMENT_REFERENCE_NAME;
-            mReferenceIndex = NO_ALIGNMENT_REFERENCE_INDEX;
-        }
-        else if (null != mHeader) {
+        if (null != mHeader) {
+            mReferenceIndex = resolveIndexFromName(referenceName, mHeader, false);
             // String.intern() is surprisingly expensive, so avoid it by looking up in sequence dictionary if possible
-            final int referenceIndex = mHeader.getSequenceIndex(referenceName);
-            if (-1 != referenceIndex) {
-                setReferenceIndex(referenceIndex); // sets reference name and index
-            }
-            else {
-                mReferenceName = referenceName.intern();
-                mReferenceIndex = null;
-            }
+            mReferenceName = null == mReferenceIndex ?
+                                    referenceName.intern() :
+                                    resolveNameFromIndex(mReferenceIndex, mHeader);
```

Comment:
Its just to maintain the interning pattern that was already in place before, and retrieve the already interned object. Not sure how much its buying us. I must make that comment more explicit because even with it I had to think about it a bit to recall why I did that.